[quant][graphmode] Supporting making submodules unique

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30037 [quant][graphmode] Supporting making submodules unique**

Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18821483](https://our.internmc.facebook.com/intern/diff/D18821483)

jerryzh168 committed:
[quant][graphmode] Supporting making submodules unique

Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  // Deduplicate module uses given the information we recorded before
+  void dedupModuleUses(const std::string& method_name) {
+    for (Value* v : use_to_rewrite_) {
+      const auto& path = value_to_path_map_[v];
```
Please use `value_to_path_map_.at(v)` instead of `value_to_path_map_[v]`. The latter one inserts an empty element if `v` was not in the map.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+    const script::Module& child_module,
+    const std::vector<std::string>& path) {
+  script::Module cur = module;
+  for (size_t i = 0; i < path.size() - 1; ++i) {
```
There should be a check for `path.size() == 0`, otherwise this loop will iterate until `UINT_MAX`.

jerryzh168 replied:
added filter for `path.size() == 0` earlier

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+        Value* iter = instance;
+        // trace back the instance to recover the path of the submodule
+        while (!hitGraphInput(iter)) {
+          Node* get_attr = iter->node();
```
If we assume that `iter` is `prim::GetAttr`, we should explicitly check/assert that.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+            use_to_rewrite_.push_back(instance);
+          }
+        } else {
+          GRAPH_DEBUG(
```
Please include some additional info into the log statements. Their goal is to save you running `gdb`, but if we don't print what graph we were analyzing and what value failed this condition, we still have to run `gdb` to figure it out.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+
+  // Analyze the code to record necessary information
+  // to dedup the module uses
+  void findModuleUses(const std::string& method_name) {
```
This needs to be private (as well as `dedupModuleUses`).

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  }
+
+  // Deduplicate module uses given the information we recorded before
+  void dedupModuleUses(const std::string& method_name) {
```
The argument is unused.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  script::Module module_;
+  // Map from value of module instance to the list of names of submodules
+  // starting from the top level module, e.g. ["sub1", "sub2", "relu"]
+  std::unordered_map<Value*, std::vector<std::string>> value_to_path_map_;
```
This might be a commonly useful feature if we generalize it a bit. E.g. we can add an API that would take a module and return you a map from `ivalue::object` to `std::vector<std::string>` representing a path to each submodule.

jerryzh168 replied:
that sounds a little bit different from what we have here, here we have a `Value` in IR and traced back in the graph to get the path and then from path we get the `ivalue::Object`. Maybe we can think about generalize this when similar use cases comes up later

ZolotukhinM replied:
With the current exact implementation yes, it's not general enough. However, locating a child module by its path from the root module seems general and I think it would've been useful here as well.

ZolotukhinM replied:
Also, might be worth mentioning that this map is just a cache for `getModuleAccessPath` calls.

jerryzh168 replied:
we do have this function in the PR :) `findChlidModule`, although this function takes a list, are you talking about an API to get the child module given a string path like "a.b.c"? maybe we can have this later if there is a need

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+        if (iter == self) {
+          value_to_path_map_[instance] = path;
+          auto m = findChildModule(module_, path);
+          if (module_set_.find(m._ivalue()) == module_set_.end()) {
```
Nit:
```
if (module_set_.insert(m._ivalue()).second) {
    use_to_rewrite_.push_back(instance);
}
```

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  // starting from the top level module, e.g. ["sub1", "sub2", "relu"]
+  std::unordered_map<Value*, std::vector<std::string>> value_to_path_map_;
+  std::unordered_set<script::ModulePtr> module_set_;
+  std::vector<Value*> use_to_rewrite_;
```
Nit: use**s**_to_rewrite

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -137,5 +137,13 @@ TORCH_API void FoldPrepackedWeightIntoModule(
+ *  creating a new module for each use of the module
+ */
+TORCH_API void DedupModuleUses(
+    script::Module& module);
```
Nit: formatting seems weird here.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+bool hitGraphInput(Value* value) {
+  bool hit_graph_input = false;
+  Graph* graph = value->owningGraph();
+  for (size_t i = 0; i < graph->inputs().size(); ++i) {
```
```
for (const auto& input : graph->inputs()) {
  if (value == input) {
    return true;
  }
}
return false;
```
or use `std::find`.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+        // Iterator to traverse back the GetAttr calls
+        Value* iter = instance;
+        // trace back the instance to recover the path of the submodule
+        while (!hitGraphInput(iter)) {
```
Can `iter` be anything other than `self`? Why a single check won't be enough here?

jerryzh168 replied:
we discussed this before, user could pass in module instance in forward and call forward on it

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,140 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+          path.push_back(get_attr->s(attr::name));
+          // trace back the chain of GetAttr
+          iter = get_attr->inputs()[0];
+        }
```
This probably can be factored out to a separate function.

ZolotukhinM reviewed (changes requested):
I think this can be cleaned up a bit, please find some comments inline.

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,149 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+ private:
+  // Analyze the code to record necessary information
+  // to dedup the module uses
+  void findModuleUses(const std::string& method_name) {
```
Nit: maybe pass a graph here instead of the name? Signature `void findModuleUses(const std::string&)` seems a bit strange compared to `void findModuleUses(Graph&)`

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,149 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+        Value* instance = n->inputs()[0];
+        // boundary_val is the value we get when we trace back
+        // the GetAttr access chain until we hit the input of graph
+        // or a node that is not prim::GetAttr
```
From the comment it seems that it's possible to find a node that 's neither input nor `prim::GetAttr`. When can this happen? Will all other logic in this pass work correctly in this case?

jerryzh168 replied:
for example when some function call returns a instance, and we call method on that? I think so, if that happens we'll stop the search there and go to the `else` branch starting L802

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,149 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  TORCH_INTERNAL_ASSERT(
+      path.size() > 0, "path must have at least one element.");
+  for (size_t i = 0; i < path.size() - 1; ++i) {
+    cur = module.attr(path[i]).toModule();
```
This loop is similar to `findChildModule`. If we change `findChildModule` to return the containing module of the module specified by the path (or the parent of the child module we're currently returning), then we can reuse the same code in these two places.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,149 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  return std::find(inputs.begin(), inputs.end(), value) != inputs.end();
+}
+
+Value* getModuleAccessPath(Value* instance, std::vector<std::string>& path) {
```
This and other functions above should probably be made private (static) methods of `ModuleUseDeduper`.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -137,5 +138,11 @@ TORCH_API void FoldPrepackedWeightIntoModule(
     const script::Module& conv_params_module);
+
+/** Deduplicate multiple uses of the same module by
+ *  creating a new module for each use of the module
```
It's probably worth adding more details in this comment: i.e. that this is recursive, it adds new submodule (a clone) to the module object **and** modifies the code of the type, what are the types of the created clones, what happens to the parameters/attributes of the cloned modules, etc.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,149 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  while (module.hasattr(child_name)) {
+    child_name = cur_name + "_" + c10::to_string(uid++);
+  }
+  module.register_module(child_name, child_module.clone());
```
This probably should use `clone_instance`.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -692,6 +695,149 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+      // change the name in GetAttr call
+      auto original_name = v->node()->s(attr::name);
+      v->node()->s_(attr::name, child_name);
+      GRAPH_DEBUG(
```
`GRAPH_UPDATE` seems more appropriate here since we're actually changing the IR.

ZolotukhinM reviewed (commented):
This is looking better! Please find some comments inline.

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -716,6 +719,150 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+
+ private:
+  // Analyze the code to record necessary information
+  // to dedup the module uses
```
Specify what information is being recorded

jamesr66a commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -716,6 +719,150 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+  // starting from the top level module, e.g. ["sub1", "sub2", "relu"]
+  // Also this is a cache of calling `getModuleAccessPath` of the value
+  std::unordered_map<Value*, std::vector<std::string>> value_to_path_map_;
+  std::unordered_set<script::ModulePtr> module_set_;
```
This variable name just restates the type (i.e. set of module). It would be better to name it for what it is semantically used for.

jamesr66a commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -716,6 +719,150 @@ c10::optional<IValue> toTwoElementIntList(Value* v) {
+    // Iterator to traverse back the GetAttr calls
+    Value* iter = instance;
+    // trace back the instance to recover the path of the submodule
+    while (!hitGraphInput(iter) && iter->node()->kind() == prim::GetAttr) {
```
Is the assumption here that if you hit a graph input, you've hit the `self` value? This is going to break for modules that are passed in as arguments, for example:

```
import torch

class Nest1(torch.nn.Module):
    def forward(self, x, my_activation):
        return my_activation(x) + my_activation(x)

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()
        self.nest1 = Nest1()

    def forward(self, x):
        return self.nest1(x, self.relu)


f = Foo()
traced = torch.jit.trace(f, (torch.rand(3, 4),))

traced._c.dump(True, False, False)

torch._C._jit_pass_dedup_module_uses(traced._c)

traced._c.dump(True, False, False)
```

(and yes, people do do this, especially for activation modules)

I'll think through a design for how we can do this correctly in the general case

jerryzh168 replied:
No, this case is taken care of, after we hit graph input we'll check if the value is equal to self, see L761

jerryzh168 replied:
we don't support this use in this PR, is there a need to support this?

jamesr66a replied:
It's not a good idea to check in this pass if it creates an ill-formed Module, as in my example above

jerryzh168 replied:
I'm not sure I understand, what do you mean?

jerryzh168 replied:
For your example there should be no effect to the module. 
I just found a bug that reverses the check of whether we should copy the use or not, it is fixed now.

jamesr66a reviewed (changes requested):
We need to do some more design work

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a reviewed (approved):
I think this is good for now. We should still keep an eye out for the edge cases not covered by this method

jerryzh168 committed:
Update on "[quant][graphmode] Supporting making submodules unique"


Summary:
Support quantization for modules with reused submodules, e.g. relu (automatically make unique)
We first do a pass on the graph to find all duplicate uses of the same module, and record the `Value`s of the
module instance, for each of these values we create a new module and change the access to that module.

Test Plan:
python test/test_jit.py
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18821483](https://our.internmc.facebook.com/intern/diff/D18821483)

[ghstack-poisoned]
